crates/telio-starcast/src/multicaster.rs

```diff
@@ -0,0 +1,281 @@
+            .map(|sock_addr| (sock_addr.ip(), sock_addr.port()))
+        {
+            (addr, dst_port)
+        } else {
+            return;
+        };
+        if let Some(mut ip_packet) = MutableIpv4Packet::new(&mut packet) {
+            ip_packet.set_destination(dst_ip);
+            ip_packet.set_checksum(0);
+            ip_packet.set_checksum(checksum(&ip_packet.to_immutable()));
+        } else {
+            return;
+        };
+        if let Some(mut udp_packet) = packet
+            .get_mut(IPV4_HEADER..)
+            .and_then(MutableUdpPacket::new)
+        {
+            udp_packet.set_destination(dst_port);
+            // Ignoring L4 checksum as it is already encapsulated in L3/L4 header
+            udp_packet.set_checksum(0x0000);
+        } else {
+            return;
+        }
+        let _ = self.packet_chan.tx.send(packet).await;
+    }
+
+    async fn handle_misc_packet(&mut self, mut packet: Vec<u8>) {
+        let (src_ip, dst_ip) = if let Some(mut ip_packet) = MutableIpv4Packet::new(&mut packet) {
+            let res = (ip_packet.get_source(), ip_packet.get_destination());
+            ip_packet.set_destination(self.vpeer_ip);
+            ip_packet.set_checksum(0);
+            ip_packet.set_checksum(checksum(&ip_packet.to_immutable()));
+            res
+        } else {
+            return;
+        };
+        if let Some(mut udp_packet) = packet
+            .get_mut(IPV4_HEADER..)
+            .and_then(MutableUdpPacket::new)
+        {
+            let natted_port = Self::map_to_port(src_ip, dst_ip, udp_packet.get_destination());
+            self.nat_map
+                .write()
+                .await
+                .entry(natted_port)
+                .or_insert_with(|| SocketAddr::new(IpAddr::V4(src_ip), udp_packet.get_source()));
+            udp_packet.set_destination(natted_port);
+            // Ignoring L4 checksum as it is already encapsulated in L3/L4 header
+            udp_packet.set_checksum(0x0000);
+        } else {
+            return;
+        }
+        let _ = self.packet_chan.tx.send(packet).await;
+    }
+
+    // TODO: Change this to a better mapper
+    fn map_to_port(src_ip: Ipv4Addr, dst_ip: Ipv4Addr, dst_port: u16) -> u16 {
+        let mut hasher = DefaultHasher::new();
+        let combined = format!("Src {} {}:{}", src_ip, dst_ip, dst_port);
+        combined.hash(&mut hasher);
+        let hash_value = hasher.finish();
+        (hash_value % 65535) as u16
+    }
+}
+
+#[async_trait]
+impl Runtime for MulticastListener {
+    const NAME: &'static str = "MulticastListener";
+
+    type Err = ();
+
+    async fn wait(&mut self) -> WaitResponse<'_, Self::Err> {
+        tokio::select! {
+            Some(packet) = self.packet_chan.rx.recv() => self.route_packet(packet, PacketOrigin::Channel).await,
+            Ok(bytes_read) = self.multicast_socket.recv(&mut self.recv_buffer) => {
+                let packet = self.recv_buffer[..bytes_read].to_vec();
```

Comment:
It would be smart to use the recv_from, and map ip to pk of other node, we will need it once extra filtering will start to happen.